libs/remote_storage/src/s3_bucket.rs

```diff
@@ -569,64 +624,138 @@ impl RemoteStorage for S3Bucket {
             delete_objects.push(obj_id);
         }
 
-        for chunk in delete_objects.chunks(MAX_KEYS_PER_DELETE) {
-            let started_at = start_measuring_requests(kind);
+        self.delete_oids(kind, &delete_objects).await
+    }
 
-            let resp = self
-                .client
-                .delete_objects()
-                .bucket(self.bucket_name.clone())
-                .delete(
-                    Delete::builder()
-                        .set_objects(Some(chunk.to_vec()))
-                        .build()?,
-                )
-                .send()
-                .await;
+    async fn delete(&self, path: &RemotePath) -> anyhow::Result<()> {
+        let paths = std::array::from_ref(path);
+        self.delete_objects(paths).await
+    }
 
-            let started_at = ScopeGuard::into_inner(started_at);
-            metrics::BUCKET_METRICS
-                .req_seconds
-                .observe_elapsed(kind, &resp, started_at);
+    async fn time_travel_recover(
+        &self,
+        prefix: Option<&RemotePath>,
+        timestamp: SystemTime,
+        done_if_after: SystemTime,
+    ) -> anyhow::Result<()> {
+        let kind = RequestKind::TimeTravel;
+        let _guard = self.permit(kind).await;
+
+        let timestamp = DateTime::from(timestamp);
+        let done_if_after = DateTime::from(done_if_after);
+
+        tracing::info!("Target time: {timestamp:?}, done_if_after {done_if_after:?}");
+
+        // get the passed prefix or if it is not set use prefix_in_bucket value
+        let prefix = prefix
+            .map(|p| self.relative_path_to_s3_object(p))
+            .or_else(|| self.prefix_in_bucket.clone());
+
+        let list = self
+            .client
+            .list_object_versions()
+            .bucket(self.bucket_name.clone())
+            .set_prefix(prefix.clone())
+            .send()
+            .await?;
```

Comment:
The object versions API is paginated and the response will contain at most 1000 keys (see [here](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/list_object_versions/builders/struct.ListObjectVersionsFluentBuilder.html#method.set_max_keys)). 

Since our objects are immutable, the dimension of versions doesn't concern me that much, but it might that we've been running for a long time and accumulated a lot of layers. Feels like it would be safer to deal with the pagination (check for [truncation](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/list_object_versions/struct.ListObjectVersionsOutput.html#method.is_truncated), etc)

Reply:
I was aware and wanted to cross the bridge when we come to it, but I can also add it now.

Reply:
That's also an option, but I'd suggest bailing out if the response is truncated. Otherwise, we wouldn't know we've crossed the threshold.

Reply:
I have added a `bail!` for the case the response is truncated. Later, we can implement processing of the longer stream.